userspace/libscap/scap_vtable.h

```diff
@@ -218,6 +218,21 @@ struct scap_vtable {
 	 */
 	int32_t (*get_stats)(struct scap_engine_handle engine, struct scap_stats *stats);
 
+	/**
+	 * @brief get engine estimate of size of statistics buffers
+	 * @return size_t of required buffer size to support each statistic
+	 */
+	size_t (*get_stats_size_hint)();
+
+	/**
+	 * @brief get engine statistics (including counters and `bpftool prog show` like stats)
+	 * @param engine wraps the pointer to the engine-specific handle
+	 * @param buf_size size of buffer to hold statistics
+	 * @param stats Pointer to a \ref scap_stats structure that will be filled with the statistics.
+	 * @return SCAP_SUCCESS or a failure code
+	 */
+	int32_t (*get_stats_v2)(struct scap_engine_handle engine, size_t buf_size, uint32_t flags, struct scap_stats_v2 *stats);
```

Comment:
Design question. Here the owner of the memory is the caller (a.k.a. libsinsp), and you are submitting an output `scap_stats_v2` struct to be filled-up by the engine, right? Why don't we remodel this function to:

```
struct scap_stats_v2 * (*get_stats_v2)(struct scap_engine_handle engine, uint32_t flags, OUT uint32_t* nstats, OUT int32_t* rc);
```

In this way, the engines will be owner of their own memory about the stats buffer. It will be easier to use for the caller by not having to prepare a big-enough buffer, thus `get_stats_size_hint` could go away as well and the whole flow simplifies. Moreover, we would avoid the extra output memory copy too by passing a buffer allocated on the engine right away.

Maybe I missed the big picture, but what do you think?

Reply:
We can do that, I have no strong opinion and I think we are still fluid in design. Anyone else having more thoughts?